### PR TITLE
add 4x4 keypad driver

### PR DESCRIPTION
**Pull request by Nerzal**
As promised some time ago, here is the PR for a 4x4 membrane keypad driver.
Solves: #212

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -166,12 +166,14 @@ endif
 	tinygo build -size short -o ./build/test.hex -target=circuitplay-express ./examples/lis2mdl/main.go
 	@md5sum ./build/test.hex
 	tinygo build -size short -o ./build/test.hex -target=feather-m0 ./examples/dht/main.go
+	@md5sum ./build/t
```
Comment (deadprogram) [inline Makefile]: Looks like the end of the line of text got cutoff here. That is why the tests are failing.

Comment (Nerzal) [reply Makefile]: next time i'll run them manually again, after resolving merge conflicts :)

**Comment by Nerzal** [comment]
Locally i still run into errors, when running make test. 

```bash
package command-line-arguments
        imports machine
        imports device/stm32: build constraints exclude all Go files in /home/tobias/.cache/tinygo/goroot-go1.16-65176e9981f9146086a1cca207bd1d47f4da8a15455d548a5d84fac12d1814d2-syscall/src/device/stm32
make: *** [Makefile:36: smoke-test] Fehler 1
```

**Comment by Nerzal** [comment]
I still run into issues locally, when running make test -> 

```
tinygo build -size short -o ./build/test.hex -target=bluepill ./examples/ds1307/sram/main.go
package command-line-arguments
        imports machine
        imports device/stm32: build constraints exclude all Go files in /home/tobias/.cache/tinygo/goroot-go1.16-65176e9981f9146086a1cca207bd1d47f4da8a15455d548a5d84fac12d1814d2-syscall/src/device/stm32
make: *** [Makefile:36: smoke-test] Fehler 1
```

File: keypad4x4/keypad4x4.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+// GetIndices returns the position of the pressed key
+func (keypad *device) GetIndices() (int, int) {
+	for rowIndex := range keypad.rows {
+		rowPin := keypad.rows[rowIndex]
+		rowPin.Low()
+
+		for columnIndex := range keypad.columns {
+			columnPin := keypad.columns[columnIndex]
```
Comment (sago35) [inline keypad4x4/keypad4x4.go:84]: It is simpler to write the following

```suggestion
		for columnIndex, columnPin := range keypad.columns {
```

File: keypad4x4/keypad4x4.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+		return 255
+	}
+
+	return keypad.mapping[row][column]
+}
+
+// GetIndices returns the position of the pressed key
+func (keypad *device) GetIndices() (int, int) {
+	for rowIndex := range keypad.rows {
+		rowPin := keypad.rows[rowIndex]
```
Comment (sago35) [inline keypad4x4/keypad4x4.go]: It is simpler to write the following

```suggestion
	for rowIndex, rowPin := range keypad.rows {
```

File: keypad4x4/keypad4x4.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+// Example:
+// 0	1	2	3
+// 4	5	6	7
+// 8	9	10	11
+// 12	13	14	15
+// returns 255 for no keyPressed
+func (keypad *device) GetKey() uint8 {
+	row, column := keypad.GetIndices()
+	if row == -1 && column == -1 {
+		return 255
```
Comment (sago35) [inline keypad4x4/keypad4x4.go]: It is better to give it some kind of name and publish it.

File: examples/keypad4x4/main.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+		15: "#",
+		16: "D",
+	}
+
+	keypadDevice := keypad4x4.NewDevice(machine.D2, machine.D3, machine.D4, machine.D5, machine.D6, machine.D7, machine.D8, machine.D9)
+	keypadDevice.Configure()
+
+	for {
+		key := keypadDevice.GetKey()
+		if key != 255 {
```
Comment (sago35) [inline examples/keypad4x4/main.go]: It is better to give it some kind of name and publish it.

**Comment by sago35** [comment]
> I still run into issues locally, when running make test ->

I was able to build it without any problems in my environment.
Do you still have problems? @Nerzal 

```
$ tinygo version
tinygo version 0.18.0-dev-303acf5 linux/amd64 (using go version go1.15.2 and LLVM version 11.0.0)

$ tinygo build -o test.hex --target arduino --size short ./examples/keypad4x4/
   code    data     bss |   flash     ram
   5778     126     649 |    5904     775
```

**Comment by Nerzal** [comment]
@sago35 yep, this example builds but when running all tests using the make test command this fails -> 

```
tinygo build -size short -o ./build/test.hex -target=bluepill ./examples/ds1307/sram/main.go
package command-line-arguments
        imports machine
        imports device/stm32: build constraints exclude all Go files in /home/tobias/.cache/tinygo/goroot-go1.16-65176e9981f9146086a1cca207bd1d47f4da8a15455d548a5d84fac12d1814d2-syscall/src/device/stm32
make: *** [Makefile:36: smoke-test] Fehler 1
```

File: keypad4x4/keypad4x4.go (hunk 4 above)
Comment (Nerzal) [reply keypad4x4/keypad4x4.go]: fixed

File: examples/keypad4x4/main.go (hunk 5 above)
Comment (Nerzal) [reply examples/keypad4x4/main.go]: fixed

**Comment by sago35** [comment]
I was able to build it.
However, I was not able to build the hifiv one.

This problem is not related to this PR, so I think it is not a problem.

```
$ tinygo version
tinygo version 0.18.0-dev-392808b windows/amd64 (using go version go1.16 and LLVM version 11.0.0)

sago3@B450GT3 MINGW64 ~/dev/src/github.com/tinygo-org/drivers (Nerzal-device/keypad)
$ tinygo build -size short -o ./build/test.hex -target=bluepill ./examples/ds1307/sram/main.go
   code    data     bss |   flash     ram
  12224      44    2224 |   12268    2268
```

error : 

```
tinygo build -size short -o ./build/test.hex -target=hifive1b ./examples/ssd1351/main.go
panic: missing core function "(*internal/task.Task).setState"

goroutine 43 [running]:
github.com/tinygo-org/tinygo/transform.Optimize(0x6af0530, 0xc0003052c0, 0x2, 0x2, 0x5, 0x0, 0x0, 0x0)
        C:/tinygo/tinygo/transform/optimizer.go:36 +0x1505
github.com/tinygo-org/tinygo/builder.compileWholeProgram(0xc000010200, 0x1a, 0xc0003052c0, 0xc0003100b0, 0xc000310160, 0x6ae3400, 0x4837da0, 0x496eee0, 0xc0013f7f60)
        C:/tinygo/tinygo/builder/build.go:423 +0x60c
github.com/tinygo-org/tinygo/builder.Build.func1(0xc00e40e6d1dac080, 0x4ee77eb5)
        C:/tinygo/tinygo/builder/build.go:97 +0x79
github.com/tinygo-org/tinygo/builder.jobWorker(0xc000069260, 0xc000069200)
        C:/tinygo/tinygo/builder/jobs.go:153 +0xf0
created by github.com/tinygo-org/tinygo/builder.runJobs
        C:/tinygo/tinygo/builder/jobs.go:73 +0xce
make: *** [Makefile:162: smoke-test] Error 2
```

**Review by sago35 — APPROVED** [review]

**Comment by deadprogram** [comment]
Now going to squash/merge. Thanks for the contribution @Nerzal and to @sago35 for the review.

**Pull request merged**